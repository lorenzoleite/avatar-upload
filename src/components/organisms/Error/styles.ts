import styled, { css } from 'styled-components';

export const StyledError = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    width: 55.3rem;
    height: 17.7rem;
    padding: 3.2rem;
    border-radius: 0.8rem;
    background-color: ${theme.color['gray.100']};
  `}
`;

export const ErrorTitle = styled.p`
  ${({ theme }) => css`
    font-family: ${theme.font.family.inter};
    font-size: ${theme.font.size.medium};
    font-weight: ${theme.font.weight.normal};
    line-height: ${theme.font.lineHeight.medium};
    color: ${theme.color['orange.100']};
  `}
`;
