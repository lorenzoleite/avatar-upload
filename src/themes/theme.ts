export const theme = {
  colors: {
    white: '#FFFFFF',

    'gray.600': '#3D485F',
    'gray.500': '#495567',
    'gray.400': '#677489',
    'gray.300': '#C7CDD3',
    'gray.200': '#C3CBD5',
    'gray.100': '#F2F5F8',

    'blue.200': '#3F80FF',
    'blue.100': '#B9D1FF',

    'orange.100': '#C64D32',
  },
  font: {
    family: 'Inter, sans-serif',
    weights: {
      normal: 400,
      medium: 500,
    },
    sizes: {
      small: '1.4rem',
      medium: '1.6rem',
    },
  },
  borders: {
    small: '0.8rem',
    medium: '1.6rem',
  },
  lineHeights: {
    small: '2.52rem',
    medium: '2.88rem',
  },
};
