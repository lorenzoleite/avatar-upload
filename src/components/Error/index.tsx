import { StyledError, FilePreview, ErrorTitle } from './styles';

import { AttentionIcon } from '../../assets/AttentionIcon';
import { CloseIcon } from '../../assets/CloseIcon';

import { Flex } from '../atoms/Flex';

import { Button } from '../Button';

type ErrorProps = {
  handleResetFile: () => void;
};

export function Error({ handleResetFile }: ErrorProps) {
  return (
    <StyledError>
      <Flex align="center">
        <FilePreview>
          <AttentionIcon />
        </FilePreview>
        <Flex direction="column" gap="0.2rem">
          <ErrorTitle>Sorry, the upload failed.</ErrorTitle>
          <Button variant="underline" onClick={() => handleResetFile()}>
            Try again
          </Button>
        </Flex>
      </Flex>
      <Flex align="flex-start">
        <Button variant="icon" onClick={() => handleResetFile()}>
          <CloseIcon />
        </Button>
      </Flex>
    </StyledError>
  );
}
