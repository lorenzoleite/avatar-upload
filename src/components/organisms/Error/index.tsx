import { StyledError, ErrorTitle } from './styles';

import { AttentionIcon } from '../../atoms/AttentionIcon';
import { CloseIcon } from '../../atoms/CloseIcon';
import { ImagePreview } from '../../molecules/ImagePreview';
import { Button } from '../../molecules/Button';
import { Flex } from '../../molecules/Flex';

type ErrorProps = {
  messageError: string;
  onClose: () => void;
};

export function Error({ messageError, onClose }: ErrorProps) {
  return (
    <StyledError>
      <Flex align="center">
        <ImagePreview>
          <AttentionIcon />
        </ImagePreview>
        <Flex direction="column" gap="0.2rem">
          <ErrorTitle>Sorry, {messageError.toLowerCase()}.</ErrorTitle>
          <Button variant="underline" onClick={() => onClose()}>
            Try again
          </Button>
        </Flex>
      </Flex>
      <Flex align="flex-start">
        <Button variant="icon" onClick={() => onClose()}>
          <CloseIcon />
        </Button>
      </Flex>
    </StyledError>
  );
}
