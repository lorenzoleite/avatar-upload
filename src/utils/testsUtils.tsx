import React from 'react';
import fetch from 'node-fetch';

import { ThemeProvider } from 'styled-components';
import { render, RenderResult } from '@testing-library/react';

import { theme } from '../themes/theme';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const imageFile = await fetch(
  'https://images.unsplash.com/photo-1661956602868-6ae368943878?ixlib=rb-4.0.3&ixid=MnwxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=234&q=16',
)
  .then(response => response.blob())
  .then(blob => new File([blob], 'image.jpg', { type: 'image/jpeg' }));
