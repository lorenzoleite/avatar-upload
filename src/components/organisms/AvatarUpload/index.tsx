import { useMemo, useState } from 'react';
import { useDropzone } from 'react-dropzone';

import { Dropzone, DropzoneTitle, DropzoneSubtitle, Image } from './styles';

import { theme } from '../../../themes/theme';

import { MediaIcon } from '../../atoms/MediaIcon';
import { ImagePreview } from '../../molecules/ImagePreview';
import { Flex } from '../../molecules/Flex';
import { Error } from '../Error';
import { Crop } from '../Crop';

type AvatarUploadProps = {};

const acceptStyle = {
  borderColor: theme.color['blue.200'],
};

const rejectStyle = {
  borderColor: theme.color['orange.100'],
};

export function AvatarUpload({}: AvatarUploadProps) {
  const [imageFile, setImageFile] = useState<File | null>(null);
  const [croppedImage, setCroppedImage] = useState<string | null>(null);
  const [hasError, setHasError] = useState<boolean>(false);
  const [messageError, setMessageError] = useState<string | null>(null);

  const { getRootProps, getInputProps, isDragAccept, isDragReject } = useDropzone({
    accept: { 'image/png': ['.png'], 'image/jpeg': ['.jpeg', '.jpg'] },
    maxFiles: 1,
    onDropAccepted: file => {
      setImageFile(file[0]);
    },
    onDropRejected: file => {
      setHasError(true);
      setMessageError(file[0].errors[0].message);
    },
  });

  const handleClose = () => {
    setImageFile(null);
    setCroppedImage(null);
    setHasError(false);
    setMessageError(null);
  };

  const handleSave = (imageUrl: string) => {
    setImageFile(null);
    setCroppedImage(imageUrl);
  };

  const style = useMemo(
    () => ({
      ...(isDragAccept && acceptStyle),
      ...(isDragReject && rejectStyle),
    }),
    [isDragAccept, isDragReject],
  );

  return (
    <Flex>
      {!imageFile && !hasError && (
        <Dropzone {...getRootProps({ style })} justify={croppedImage ? 'flex-start' : 'center'} role="dropzone">
          <input {...getInputProps()} data-testid="inputDropzone" />
          <Flex align="center">
            {croppedImage && (
              <ImagePreview marginRight="5rem">
                <Image src={croppedImage} alt="Image Preview" />
              </ImagePreview>
            )}
            <Flex direction="column" align="center">
              <Flex align="center" gap="1.2rem">
                <MediaIcon />
                <DropzoneTitle>Organization Logo</DropzoneTitle>
              </Flex>
              <DropzoneSubtitle>Drop the image here or click to browse.</DropzoneSubtitle>
            </Flex>
          </Flex>
        </Dropzone>
      )}
      {hasError && messageError && <Error messageError={messageError} onClose={handleClose} />}
      {imageFile && <Crop imageFile={imageFile} onSave={handleSave} onClose={handleClose} />}
    </Flex>
  );
}
