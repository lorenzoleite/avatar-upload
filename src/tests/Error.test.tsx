import { screen } from '@testing-library/react';
import { describe, it, vi } from 'vitest';

import { Error } from '../components/organisms/Error';
import { renderWithTheme } from '../utils/testsUtils';
import { theme } from '../themes/theme';

describe('Error', () => {
  it('should render the Error', () => {
    const onClose = vi.fn();
    const { container } = renderWithTheme(<Error messageError="Sorry, the upload failed." onClose={onClose} />);

    expect(screen.getByTestId(/attentionIcon/i)).toBeInTheDocument();
    expect(screen.getByText(/Sorry,/i)).toHaveStyle({
      color: theme.color['orange.100'],
    });
    expect(screen.getByRole('button', { name: /Try again/i })).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
