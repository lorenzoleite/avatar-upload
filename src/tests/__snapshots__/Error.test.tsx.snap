// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error > should render the Error 1`] = `
<div
  class="sc-fmSAUk hiXeEW"
>
  <div
    class="sc-dKfzgJ cOLHTs"
  >
    <div
      class="sc-blLsxD klCXxN"
    >
      <svg
        data-testid="attentionIcon"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          clip-path="url(#clip0_1_34)"
        >
          <path
            clip-rule="evenodd"
            d="M9.988-.012c-5.525 0-10 4.475-10 10 0 5.524 4.474 10 10 10 5.524 0 10-4.476 10-10 0-5.525-4.476-10-10-10zm1.25 16.25h-2.5v-2.5h2.5v2.5zm0-3.75h-2.5v-8.75h2.5v8.75z"
            fill="#fff"
            fill-rule="evenodd"
          />
        </g>
        <defs>
          <clippath
            id="clip0_1_34"
          >
            <path
              d="M0 0H20V20H0z"
              fill="#fff"
            />
          </clippath>
        </defs>
      </svg>
    </div>
    <div
      class="sc-dKfzgJ iygmQv"
      direction="column"
    >
      <p
        class="sc-fGFwAa GHXaE"
      >
        Sorry, 
        sorry, the upload failed.
        .
      </p>
      <button
        class="sc-ieZDjg crovzC"
      >
        Try again
      </button>
    </div>
  </div>
  <div
    class="sc-dKfzgJ klmaLt"
  >
    <button
      class="sc-ieZDjg jvvPRo"
    >
      <svg
        data-testid="closeIcon"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M11.762 10l4.113-4.112c.238-.226.375-.538.375-.888a1.254 1.254 0 00-2.137-.888L10 8.239 5.888 4.113a1.254 1.254 0 00-1.775 1.775L8.238 10l-4.113 4.113A1.207 1.207 0 003.75 15a1.254 1.254 0 002.138.887L10 11.762l4.113 4.113c.225.238.537.375.887.375a1.254 1.254 0 00.887-2.137L11.762 10z"
          fill="#677489"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
